streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -225,13 +241,37 @@ void addGraphNode(final Collection<StreamsGraphNode> parents, final StreamsGraph
         }
     }
 
-    void maybeAddNodeForOptimizationMetadata(final StreamsGraphNode node) {
-        node.setId(nodeIdCounter.getAndIncrement());
+    private void maybeAddNodeForOptimizationMetadata(final StreamsGraphNode node) {
+        node.setBuildPriority(buildPriorityIndex.getAndIncrement());
+
+        if (node.parentNodes().isEmpty() && !node.nodeName().equals(TOPOLOGY_ROOT)) {
+            throw new IllegalStateException(
+                "Nodes should not have a null parent node.  Name: " + node.nodeName() + " Type: "
+                + node.getClass().getSimpleName());
+        }
+
+        if (node instanceof TableSourceNode && !((TableSourceNode) node).isGlobalKTable()) {
+            tableSourceNodes.add((TableSourceNode) node);
+        } else if (node.isKeyChangingOperation()) {
+            keyChangingOperationsToOptimizableRepartitionNodes.put(node, new HashSet<>());
+        } else if (node instanceof OptimizableRepartitionNode) {
+            final StreamsGraphNode parentNode = getKeyChangingParentNode(node);
+            if (parentNode != null) {
+                keyChangingOperationsToOptimizableRepartitionNodes.get(parentNode).add((OptimizableRepartitionNode) node);
+            }
+        }
     }
 
+    // use this method for testing only
     public void buildAndOptimizeTopology() {
+        buildAndOptimizeTopology(null);
+    }
+
+    public void buildAndOptimizeTopology(final Properties props) {
 
-        final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, nodeIdComparator);
+        maybePerformOptimizations(props);
+
+        final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, buildPriorityComparator);
```

Comment:
nit:
This could be simplified:
``` java
final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = 
    new PriorityQueue<>(5, Comparator.comparing(StreamsGraphNode::buildPriority));
```

Reply:
ack